fmf/plugins/pytest/tmt_semantic.py

```diff
@@ -0,0 +1,273 @@
+import inspect
+import shlex
+import unittest
+
+from fmf.utils import GeneralError
+
+TEST_METHOD_PREFIX = "test"
+FMF_ATTR_PREFIX = "__fmf_"
+FMF_POSTFIX_MARKS = ("+", "-", "")
+SUMMARY_KEY = "summary"
+DESCRIPTION_KEY = "description"
+ENVIRONMENT_KEY = "environment"
+_ = shlex
+
+TMT_ATTRIBUTES = {
+    SUMMARY_KEY: str,
+    DESCRIPTION_KEY: str,
+    "order": int,
+    "adjust": (
+        list,
+        dict,
+        ),
+    "tag": (
+        list,
+        str,
+        ),
+    "link": (list, str, dict),
+    "duration": str,
+    "tier": str,
+    "component": (
+        list,
+        str,
+        ),
+    "require": (
+        list,
+        str,
+        dict,
+        ),
+    "test": (str,),
+    "framework": (str,),
+    ENVIRONMENT_KEY: (
+        dict,
+        str,
+        ),
+    "path": (str,),
+    "enabled": (bool,),
+}
+
+
+def fmf_prefixed_name(name):
+    return FMF_ATTR_PREFIX + name
+
+
+class __TMTMeta(type):
+    @staticmethod
+    def _set_fn(name, base_type=None):
+        if name not in TMT_ATTRIBUTES:
+            raise GeneralError(
+                "fmf decorator {} not found in {}".format(
+                    name, TMT_ATTRIBUTES.keys()))
+
+        def inner(*args, post_mark=""):
+            return generic_metadata_setter(
+                fmf_prefixed_name(name),
+                args,
+                base_type=base_type or TMT_ATTRIBUTES[name],
+                post_mark=post_mark,
+                )
+
+        return inner
+
+    def __getattr__(cls, name):
+        return cls._set_fn(name)
+
+
+class TMT(metaclass=__TMTMeta):
```

Comment:
One thing I can say I disagree with right away: this approach taints fmf with a custom code to make a single library user happy and makes fmf do things that are IMO well beyond its scope.

First, it makes adding and changing tmt specifications harder, adding a new place that needs to be updated when specification changes. But, more importantly: I don't think it's up to fmf to inspect the input or validate it in any way.

Is fmf validating or converting tmt keys when reading them from a genuine fmf file? I don't think that's the case, and it should not happen here either. If I put `tag: false` into my `main.fmf` of a test, fmf library will not complain, it will not touch the value, it will not check against a tmt specification schema, it will read it, put it into a fmf tree, and it will deliver the key and its value to tmt. End of story. It's tmt who's responsible for validating the input and reporting its type is invalid and unsupported.

Would it be possible to make the pytest plugin as "dumb" as the actual fmf processing, i.e. treating the input values as opaque blobs of key & value bits, and stacking them into a final fmf tree? I bet it would be possible with some clever Python magic, either have an open decorator accepting all kinds of kwargs or some dynamic decorators - but I **strongly** believe the openness here is a **good thing**, not a bug - fmf should not care about key names or value types, it does not care about them when reading them from `*.fmf` files.

```python
# Probably the simples, but less readable:
@fmf(
    tag=['foo', 'bar'],
    adjust=[
        {...}
...)

# Or maybe allow YAML-in-a-string and parse it as if read from an fmf file?
@fmf(
"""
author: foo@bar.baz
tag:
  - foo
  - bar
adjust:
...
""")

# Or with some __getattr__ magic, this could be possible by creating dynamic decorators aware of their name, i.e. the key:
@fmf.author('foo@bar.baz')
@fmf.tag(['foo', 'bar'])
...
```

tmt is and should remain the sole owner of the semantics. Just like any other tool using fmf to read data. It's not up to fmf to judge, nobody asked fmf's opinion on keys from `*.fmf` files, why would it matter here?

Should a competing test case management tool appear, build on top of fmf too, but with a different set of keys, would fmf learn its semantics as well? And when another two appear, these brand new and cool `mmt` and `ttm` tools, will fmf again learn about their semantics? I don't think that's a good road to take :)

Reply:
you are right, semantics, could be part of another tool. I've Implemented it as one module, because in this case it is all in one solution. I've wanted to show complete solution. My idea was that in FMF there will be just support for plugin, and implementation of plugins will be separate project or part of TMT. But plugin solution alone is hard to test itself. so that complet solution also proves how it works and it also show usecase in TMT or whatever else. I'm familiar to split it. But we have to agreed on the solution for plugin. and improve this part to have it testable as much as possible and when we found it is good enough then I'll split it.